// osu-api-extended/dist/api/v2/assets_backgrounds.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.assets_backgrounds = void 0;
const handleErrors_1 = require("../../utility/handleErrors");
const request_1 = require("../../utility/request");
const assets_backgrounds = (params, addons) => {
    if ((params === null || params === void 0 ? void 0 : params.type) == 'seasonal') {
        return (0, request_1.request)(`https://osu.ppy.sh/api/v2/seasonal-backgrounds`, {
            method: 'GET',
            addons,
        });
    }
    ;
    if ((params === null || params === void 0 ? void 0 : params.type) == 'beatmapset') {
        if (params.set_id == null) {
            return (0, handleErrors_1.handleErrors)(new Error('Specify beatmapset id'));
        }
        ;
        return {
            cover: `https://assets.ppy.sh/beatmaps/${params.set_id}/covers/cover.jpg`,
            "cover@2x": `https://assets.ppy.sh/beatmaps/${params.set_id}/covers/cover@2x.jpg`,
            card: `https://assets.ppy.sh/beatmaps/${params.set_id}/covers/card.jpg`,
            "card@2x": `https://assets.ppy.sh/beatmaps/${params.set_id}/covers/card@2x.jpg`,
            list: `https://assets.ppy.sh/beatmaps/${params.set_id}/covers/list.jpg`,
            "list@2x": `https://assets.ppy.sh/beatmaps/${params.set_id}/covers/list@2x.jpg`,
            slimcover: `https://assets.ppy.sh/beatmaps/${params.set_id}/covers/slimcover.jpg`,
            "slimcover@2x": `https://assets.ppy.sh/beatmaps/${params.set_id}/covers/slimcover@2x.jpg`,
            raw: `https://assets.ppy.sh/beatmaps/${params.set_id}/covers/raw.jpg`,
            fullsize: `https://assets.ppy.sh/beatmaps/${params.set_id}/covers/fullsize.jpg`,
        };
    }
    ;
    return (0, handleErrors_1.handleErrors)(new Error(`Unsupported type: ${params === null || params === void 0 ? void 0 : params.type}`));
};
exports.assets_backgrounds = assets_backgrounds;
