// osu-api-extended/dist/api/v2/forums_topics_list.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.forums_topics_list = void 0;
const request_1 = require("../../utility/request");
const handleErrors_1 = require("../../utility/handleErrors");
const forums_topics_list = (params, addons) => __awaiter(void 0, void 0, void 0, function* () {
    if ((params === null || params === void 0 ? void 0 : params.id) == null) {
        return (0, handleErrors_1.handleErrors)(new Error('Specify forum id'));
    }
    ;
    const data = yield (0, request_1.request)(`https://osu.ppy.sh/api/v2/forums/topics`, {
        method: 'GET',
        params: {
            forum_id: params.id,
            sort: params.sort,
            limit: params.limit,
            cursor_string: params.cursor_string,
        },
        addons,
    });
    if (data.error)
        return (0, handleErrors_1.handleErrors)(new Error(data.error));
    return data;
});
exports.forums_topics_list = forums_topics_list;
