// osu-api-extended/dist/api/v2/news_list.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.news_list = void 0;
const handleErrors_1 = require("../../utility/handleErrors");
const request_1 = require("../../utility/request");
const news_list = (params = {}, addons) => __awaiter(void 0, void 0, void 0, function* () {
    const object = {
        year: params === null || params === void 0 ? void 0 : params.from_year,
        limit: params === null || params === void 0 ? void 0 : params.limit,
        cursor_string: params === null || params === void 0 ? void 0 : params.cursor_string,
    };
    const data = yield (0, request_1.request)(`https://osu.ppy.sh/api/v2/news`, {
        method: 'GET',
        params: object,
        addons,
    });
    if (data.error)
        return (0, handleErrors_1.handleErrors)(new Error(data.error));
    return data;
});
exports.news_list = news_list;
