// osu-api-extended/dist/api/v1/routes/match.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.description = void 0;
const handleErrors_1 = require("../../../utility/handleErrors");
const request_1 = require("../../../utility/request");
const match_1 = __importDefault(require("../form/match"));
exports.description = {
    auth: 1,
    title: __filename,
    method: 'GET',
    description: 'Return match details',
    params: [
        {
            type: 'string',
            name: 'id',
            optional: false,
            description: 'id of the user',
        },
    ],
};
;
const name = (id) => __awaiter(void 0, void 0, void 0, function* () {
    const params = { mp: id };
    const data = yield (0, request_1.request)(`https://osu.ppy.sh/api/get_match`, {
        method: 'GET',
        params: params,
    });
    if (data.error)
        return (0, handleErrors_1.handleErrors)(new Error(data.error));
    if (data.match == 0 || !data.match) {
        return (0, handleErrors_1.handleErrors)(new Error('Match not found'));
    }
    ;
    const format = (0, match_1.default)(data);
    return format;
});
exports.default = name;
