// osu-api-extended@3.1.4/dist/types/enums.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CountryCodes = exports.CountryNames = exports.ModsOrder = exports.ModsEnumBitwise = exports.ModsEnum = exports.ModsCodes = exports.Languages_enum = exports.Genres_enum = exports.GamemodeEnum = void 0;
var GamemodeEnum;
(function (GamemodeEnum) {
    GamemodeEnum[GamemodeEnum["osu"] = 0] = "osu";
    GamemodeEnum[GamemodeEnum["taiko"] = 1] = "taiko";
    GamemodeEnum[GamemodeEnum["fruits"] = 2] = "fruits";
    GamemodeEnum[GamemodeEnum["mania"] = 3] = "mania";
})(GamemodeEnum = exports.GamemodeEnum || (exports.GamemodeEnum = {}));
;
var Genres_enum;
(function (Genres_enum) {
    Genres_enum[Genres_enum["Unspecified"] = 0] = "Unspecified";
    Genres_enum[Genres_enum["Video Game"] = 1] = "Video Game";
    Genres_enum[Genres_enum["Anime"] = 2] = "Anime";
    Genres_enum[Genres_enum["Rock"] = 3] = "Rock";
    Genres_enum[Genres_enum["Pop"] = 4] = "Pop";
    Genres_enum[Genres_enum["Other"] = 5] = "Other";
    Genres_enum[Genres_enum["Novelty"] = 6] = "Novelty";
    Genres_enum[Genres_enum["Hip Hop"] = 7] = "Hip Hop";
    Genres_enum[Genres_enum["Electronic"] = 8] = "Electronic";
    Genres_enum[Genres_enum["Metal"] = 9] = "Metal";
    Genres_enum[Genres_enum["Classical"] = 10] = "Classical";
    Genres_enum[Genres_enum["Folk"] = 11] = "Folk";
    Genres_enum[Genres_enum["Jazz"] = 12] = "Jazz";
})(Genres_enum = exports.Genres_enum || (exports.Genres_enum = {}));
;
var Languages_enum;
(function (Languages_enum) {
    Languages_enum[Languages_enum["Any"] = 0] = "Any";
    Languages_enum[Languages_enum["English"] = 1] = "English";
    Languages_enum[Languages_enum["Chinese"] = 2] = "Chinese";
    Languages_enum[Languages_enum["French"] = 3] = "French";
    Languages_enum[Languages_enum["German"] = 4] = "German";
    Languages_enum[Languages_enum["Italian"] = 5] = "Italian";
    Languages_enum[Languages_enum["Japanese"] = 6] = "Japanese";
    Languages_enum[Languages_enum["Korean"] = 7] = "Korean";
    Languages_enum[Languages_enum["Spanish"] = 8] = "Spanish";
    Languages_enum[Languages_enum["Swedish"] = 9] = "Swedish";
    Languages_enum[Languages_enum["Russian"] = 10] = "Russian";
    Languages_enum[Languages_enum["Polish"] = 11] = "Polish";
    Languages_enum[Languages_enum["Instrumental"] = 12] = "Instrumental";
    Languages_enum[Languages_enum["Unspecified"] = 13] = "Unspecified";
    Languages_enum[Languages_enum["Other"] = 14] = "Other";
})(Languages_enum = exports.Languages_enum || (exports.Languages_enum = {}));
;
exports.ModsCodes = {
    1: 'NF',
    2: 'EZ',
    4: 'TD',
    8: 'HD',
    16: 'HR',
    32: 'SD',
    64: 'DT',
    128: 'RX',
    256: 'HT',
    576: 'NC',
    1024: 'FL',
    2048: 'AT',
    4096: 'SO',
    8192: 'AP',
    16416: 'PF',
    32768: '4K',
    65536: '5K',
    131072: '6K',
    262144: '7K',
    524288: '8K',
    1048576: 'FI',
    2097152: 'RD',
    4194304: 'LM',
    8388608: 'Target',
    16777216: '9K',
    33554432: 'KeyCoop',
    67108864: '1K',
    134217728: '3K',
    268435456: '2K',
    536870912: 'ScoreV2',
    1073741824: 'MR',
};
var ModsEnum;
(function (ModsEnum) {
    ModsEnum[ModsEnum["NF"] = 1] = "NF";
    ModsEnum[ModsEnum["EZ"] = 2] = "EZ";
    ModsEnum[ModsEnum["TD"] = 4] = "TD";
    ModsEnum[ModsEnum["HD"] = 8] = "HD";
    ModsEnum[ModsEnum["HR"] = 16] = "HR";
    ModsEnum[ModsEnum["SD"] = 32] = "SD";
    ModsEnum[ModsEnum["DT"] = 64] = "DT";
    ModsEnum[ModsEnum["RX"] = 128] = "RX";
    ModsEnum[ModsEnum["HT"] = 256] = "HT";
    ModsEnum[ModsEnum["NC"] = 576] = "NC";
    ModsEnum[ModsEnum["FL"] = 1024] = "FL";
    ModsEnum[ModsEnum["AT"] = 2048] = "AT";
    ModsEnum[ModsEnum["SO"] = 4096] = "SO";
    ModsEnum[ModsEnum["AP"] = 8192] = "AP";
    ModsEnum[ModsEnum["PF"] = 16416] = "PF";
    ModsEnum[ModsEnum["4K"] = 32768] = "4K";
    ModsEnum[ModsEnum["5K"] = 65536] = "5K";
    ModsEnum[ModsEnum["6K"] = 131072] = "6K";
    ModsEnum[ModsEnum["7K"] = 262144] = "7K";
    ModsEnum[ModsEnum["8K"] = 524288] = "8K";
    ModsEnum[ModsEnum["FI"] = 1048576] = "FI";
    ModsEnum[ModsEnum["RD"] = 2097152] = "RD";
    ModsEnum[ModsEnum["LM"] = 4194304] = "LM";
    ModsEnum[ModsEnum["Target"] = 8388608] = "Target";
    ModsEnum[ModsEnum["9K"] = 16777216] = "9K";
    ModsEnum[ModsEnum["KeyCoop"] = 33554432] = "KeyCoop";
    ModsEnum[ModsEnum["1K"] = 67108864] = "1K";
    ModsEnum[ModsEnum["3K"] = 134217728] = "3K";
    ModsEnum[ModsEnum["2K"] = 268435456] = "2K";
    ModsEnum[ModsEnum["ScoreV2"] = 536870912] = "ScoreV2";
    ModsEnum[ModsEnum["MR"] = 1073741824] = "MR";
})(ModsEnum = exports.ModsEnum || (exports.ModsEnum = {}));
;
var ModsEnumBitwise;
(function (ModsEnumBitwise) {
    ModsEnumBitwise[ModsEnumBitwise["Nomod"] = 0] = "Nomod";
    ModsEnumBitwise[ModsEnumBitwise["NoFail"] = 1] = "NoFail";
    ModsEnumBitwise[ModsEnumBitwise["Easy"] = 2] = "Easy";
    ModsEnumBitwise[ModsEnumBitwise["TouchDevice"] = 4] = "TouchDevice";
    ModsEnumBitwise[ModsEnumBitwise["Hidden"] = 8] = "Hidden";
    ModsEnumBitwise[ModsEnumBitwise["HardRock"] = 16] = "HardRock";
    ModsEnumBitwise[ModsEnumBitwise["SuddenDeath"] = 32] = "SuddenDeath";
    ModsEnumBitwise[ModsEnumBitwise["DoubleTime"] = 64] = "DoubleTime";
    ModsEnumBitwise[ModsEnumBitwise["Relax"] = 128] = "Relax";
    ModsEnumBitwise[ModsEnumBitwise["HalfTime"] = 256] = "HalfTime";
    ModsEnumBitwise[ModsEnumBitwise["Nightcore"] = 512] = "Nightcore";
    ModsEnumBitwise[ModsEnumBitwise["Flashlight"] = 1024] = "Flashlight";
    ModsEnumBitwise[ModsEnumBitwise["Autoplay"] = 2048] = "Autoplay";
    ModsEnumBitwise[ModsEnumBitwise["SpunOut"] = 4096] = "SpunOut";
    ModsEnumBitwise[ModsEnumBitwise["Relax2"] = 8192] = "Relax2";
    ModsEnumBitwise[ModsEnumBitwise["Perfect"] = 16384] = "Perfect";
    ModsEnumBitwise[ModsEnumBitwise["Key4"] = 32768] = "Key4";
    ModsEnumBitwise[ModsEnumBitwise["Key5"] = 65536] = "Key5";
    ModsEnumBitwise[ModsEnumBitwise["Key6"] = 131072] = "Key6";
    ModsEnumBitwise[ModsEnumBitwise["Key7"] = 262144] = "Key7";
    ModsEnumBitwise[ModsEnumBitwise["Key8"] = 524288] = "Key8";
    ModsEnumBitwise[ModsEnumBitwise["FadeIn"] = 1048576] = "FadeIn";
    ModsEnumBitwise[ModsEnumBitwise["Random"] = 2097152] = "Random";
    ModsEnumBitwise[ModsEnumBitwise["Cinema"] = 4194304] = "Cinema";
    ModsEnumBitwise[ModsEnumBitwise["Target"] = 8388608] = "Target";
    ModsEnumBitwise[ModsEnumBitwise["Key9"] = 16777216] = "Key9";
    ModsEnumBitwise[ModsEnumBitwise["Key10"] = 33554432] = "Key10";
    ModsEnumBitwise[ModsEnumBitwise["Key1"] = 67108864] = "Key1";
    ModsEnumBitwise[ModsEnumBitwise["Key3"] = 134217728] = "Key3";
    ModsEnumBitwise[ModsEnumBitwise["Key2"] = 268435456] = "Key2";
    ModsEnumBitwise[ModsEnumBitwise["LastMod"] = 536870912] = "LastMod";
    ModsEnumBitwise[ModsEnumBitwise["keyMod"] = 521109504] = "keyMod";
    ModsEnumBitwise[ModsEnumBitwise["KeyModUnranked"] = 520093696] = "KeyModUnranked";
    ModsEnumBitwise[ModsEnumBitwise["FreeModAllowed"] = 522171579] = "FreeModAllowed";
    ModsEnumBitwise[ModsEnumBitwise["ScoreIncreaseMods"] = 1049688] = "ScoreIncreaseMods";
})(ModsEnumBitwise = exports.ModsEnumBitwise || (exports.ModsEnumBitwise = {}));
;
exports.ModsOrder = {
    nf: 0,
    ez: 1,
    hd: 2,
    dt: 3,
    nc: 3,
    ht: 3,
    hr: 4,
    so: 5,
    sd: 5,
    pf: 5,
    fl: 6,
    td: 7,
};
exports.CountryNames = {
    US: "United States",
    RU: "Russian Federation",
    DE: "Germany",
    PL: "Poland",
    FR: "France",
    JP: "Japan",
    CA: "Canada",
    BR: "Brazil",
    GB: "United Kingdom",
    TW: "Taiwan",
    KR: "South Korea",
    CN: "China",
    AU: "Australia",
    ID: "Indonesia",
    UA: "Ukraine",
    PH: "Philippines",
    CL: "Chile",
    FI: "Finland",
    AR: "Argentina",
    NL: "Netherlands",
    SE: "Sweden",
    SG: "Singapore",
    MX: "Mexico",
    MY: "Malaysia",
    ES: "Spain",
    IT: "Italy",
    HK: "Hong Kong",
    TH: "Thailand",
    VN: "Vietnam",
    NO: "Norway",
    CZ: "Czech Republic",
    TR: "Turkey",
    BY: "Belarus",
    AT: "Austria",
    BE: "Belgium",
    PT: "Portugal",
    RO: "Romania",
    HU: "Hungary",
    DK: "Denmark",
    LT: "Lithuania",
    KZ: "Kazakhstan",
    NZ: "New Zealand",
    PE: "Peru",
    CH: "Switzerland",
    CO: "Colombia",
    IL: "Israel",
    EE: "Estonia",
    BG: "Bulgaria",
    SK: "Slovakia",
    GR: "Greece",
    LV: "Latvia",
    VE: "Venezuela",
    RS: "Serbia",
    IE: "Ireland",
    HR: "Croatia",
    SA: "Saudi Arabia",
    UY: "Uruguay",
    ZA: "South Africa",
    AE: "United Arab Emirates",
    SI: "Slovenia",
    IN: "India",
    EC: "Ecuador",
    MA: "Morocco",
    CR: "Costa Rica",
    MD: "Moldova",
    DO: "Dominican Republic",
    BN: "Brunei",
    EG: "Egypt",
    RE: "Reunion",
    MO: "Macau",
    TN: "Tunisia",
    PA: "Panama",
    DZ: "Algeria",
    MN: "Mongolia",
    PY: "Paraguay",
    KW: "Kuwait",
    PR: "Puerto Rico",
    GE: "Georgia",
    BO: "Bolivia",
    SV: "El Salvador",
    QA: "Qatar",
    GT: "Guatemala",
    LU: "Luxembourg",
    UZ: "Uzbekistan",
    KG: "Kyrgyzstan",
    MK: "North Macedonia",
    KH: "Cambodia",
    BA: "Bosnia and Herzegovina",
    IS: "Iceland",
    JO: "Jordan",
    PK: "Pakistan",
    TT: "Trinidad and Tobago",
    CY: "Cyprus",
    HN: "Honduras",
    NI: "Nicaragua",
    BH: "Bahrain",
    MV: "Maldives",
    BD: "Bangladesh",
    PF: "French Polynesia",
    LB: "Lebanon",
    NP: "Nepal",
    IQ: "Iraq",
    GU: "Guam",
    AZ: "Azerbaijan",
    AL: "Albania",
    MT: "Malta",
    MM: "Myanmar",
    NC: "New Caledonia",
    OM: "Oman",
    IR: "Iran, Islamic Republic of",
    AM: "Armenia",
    GP: "Guadeloupe",
    MQ: "Martinique",
    JM: "Jamaica",
    LA: "Lao People&#039;s Democratic Republic",
    LK: "Sri Lanka",
    PS: "Palestinian Territory Occupied",
    MU: "Mauritius",
    SY: "Syrian Arab Republic",
    ME: "Montenegro",
    FO: "Faroe Islands",
    JE: "Jersey",
    GF: "French Guiana",
    MP: "Northern Mariana Islands",
    IM: "Isle of Man",
    SR: "Suriname",
    BZ: "Belize",
    BB: "Barbados",
    AW: "Aruba",
    LY: "Libya",
    AX: "Aland Islands",
    GG: "Guernsey",
    MG: "Madagascar",
    SD: "Sudan",
    LI: "Liechtenstein",
    BS: "Bahamas",
    KE: "Kenya",
    GL: "Greenland",
    BM: "Bermuda",
    EU: "Europe",
    CI: "Cote D&#039;Ivoire",
    GY: "Guyana",
    GI: "Gibraltar",
    VI: "Virgin Islands, U.S.",
    AD: "Andorra",
    LC: "Saint Lucia",
    TJ: "Tajikistan",
    AG: "Antigua and Barbuda",
    SN: "Senegal",
    PM: "Saint Pierre and Miquelon",
    CK: "Cook Islands",
    AQ: "Antarctica",
    CF: "Central African Republic",
    LR: "Liberia",
    ER: "Eritrea",
};
exports.CountryCodes = {
    "United States": 'US',
    "Russian Federation": 'RU',
    "Germany": 'DE',
    "Poland": 'PL',
    "France": 'FR',
    "Japan": 'JP',
    "Canada": 'CA',
    "Brazil": 'BR',
    "United Kingdom": 'GB',
    "Taiwan": 'TW',
    "South Korea": 'KR',
    "China": 'CN',
    "Australia": 'AU',
    "Indonesia": 'ID',
    "Ukraine": 'UA',
    "Philippines": 'PH',
    "Chile": 'CL',
    "Finland": 'FI',
    "Argentina": 'AR',
    "Netherlands": 'NL',
    "Sweden": 'SE',
    "Singapore": 'SG',
    "Mexico": 'MX',
    "Malaysia": 'MY',
    "Spain": 'ES',
    "Italy": 'IT',
    "Hong Kong": 'HK',
    "Thailand": 'TH',
    "Vietnam": 'VN',
    "Norway": 'NO',
    "Czech Republic": 'CZ',
    "Turkey": 'TR',
    "Belarus": 'BY',
    "Austria": 'AT',
    "Belgium": 'BE',
    "Portugal": 'PT',
    "Romania": 'RO',
    "Hungary": 'HU',
    "Denmark": 'DK',
    "Lithuania": 'LT',
    "Kazakhstan": 'KZ',
    "New Zealand": 'NZ',
    "Peru": 'PE',
    "Switzerland": 'CH',
    "Colombia": 'CO',
    "Israel": 'IL',
    "Estonia": 'EE',
    "Bulgaria": 'BG',
    "Slovakia": 'SK',
    "Greece": 'GR',
    "Latvia": 'LV',
    "Venezuela": 'VE',
    "Serbia": 'RS',
    "Ireland": 'IE',
    "Croatia": 'HR',
    "Saudi Arabia": 'SA',
    "Uruguay": 'UY',
    "South Africa": 'ZA',
    "United Arab Emirates": 'AE',
    "Slovenia": 'SI',
    "India": 'IN',
    "Ecuador": 'EC',
    "Morocco": 'MA',
    "Costa Rica": 'CR',
    "Moldova": 'MD',
    "Dominican Republic": 'DO',
    "Brunei": 'BN',
    "Egypt": 'EG',
    "Reunion": 'RE',
    "Macau": 'MO',
    "Tunisia": 'TN',
    "Panama": 'PA',
    "Algeria": 'DZ',
    "Mongolia": 'MN',
    "Paraguay": 'PY',
    "Kuwait": 'KW',
    "Puerto Rico": 'PR',
    "Georgia": 'GE',
    "Bolivia": 'BO',
    "El Salvador": 'SV',
    "Qatar": 'QA',
    "Guatemala": 'GT',
    "Luxembourg": 'LU',
    "Uzbekistan": 'UZ',
    "Kyrgyzstan": 'KG',
    "North Macedonia": 'MK',
    "Cambodia": 'KH',
    "Bosnia and Herzegovina": 'BA',
    "Iceland": 'IS',
    "Jordan": 'JO',
    "Pakistan": 'PK',
    "Trinidad and Tobago": 'TT',
    "Cyprus": 'CY',
    "Honduras": 'HN',
    "Nicaragua": 'NI',
    "Bahrain": 'BH',
    "Maldives": 'MV',
    "Bangladesh": 'BD',
    "French Polynesia": 'PF',
    "Lebanon": 'LB',
    "Nepal": 'NP',
    "Iraq": 'IQ',
    "Guam": 'GU',
    "Azerbaijan": 'AZ',
    "Albania": 'AL',
    "Malta": 'MT',
    "Myanmar": 'MM',
    "New Caledonia": 'NC',
    "Oman": 'OM',
    "Iran, Islamic Republic of": 'IR',
    "Armenia": 'AM',
    "Guadeloupe": 'GP',
    "Martinique": 'MQ',
    "Jamaica": 'JM',
    "Lao People&#039;s Democratic Republic": 'LA',
    "Sri Lanka": 'LK',
    "Palestinian Territory Occupied": 'PS',
    "Mauritius": 'MU',
    "Syrian Arab Republic": 'SY',
    "Montenegro": 'ME',
    "Faroe Islands": 'FO',
    "Jersey": 'JE',
    "French Guiana": 'GF',
    "Northern Mariana Islands": 'MP',
    "Isle of Man": 'IM',
    "Suriname": 'SR',
    "Belize": 'BZ',
    "Barbados": 'BB',
    "Aruba": 'AW',
    "Libya": 'LY',
    "Aland Islands": 'AX',
    "Guernsey": 'GG',
    "Madagascar": 'MG',
    "Sudan": 'SD',
    "Liechtenstein": 'LI',
    "Bahamas": 'BS',
    "Kenya": 'KE',
    "Greenland": 'GL',
    "Bermuda": 'BM',
    "Europe": 'EU',
    "Cote D &#039; Ivoire": 'CI',
    "Guyana": 'GY',
    "Gibraltar": 'GI',
    "Virgin Islands, U.S.": 'VI',
    "Andorra": 'AD',
    "Saint Lucia": 'LC',
    "Tajikistan": 'TJ',
    "Antigua and Barbuda": 'AG',
    "Senegal": 'SN',
    "Saint Pierre and Miquelon": 'PM',
    "Cook Islands": 'CK',
    "Antarctica": 'AQ',
    "Central African Republic": 'CF',
    "Liberia": 'LR',
    "Eritrea": 'ER',
};
