// osu-api-extended/dist/api/v1/routes/user/scores-category.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.description = void 0;
const handleErrors_1 = require("../../../../utility/handleErrors");
const request_1 = require("../../../../utility/request");
const best_1 = __importDefault(require("../../form/user/best"));
const recent_1 = __importDefault(require("../../form/user/recent"));
const _mode = [
    'osu',
    'taiko',
    'fruits',
    'mania',
];
exports.description = {
    auth: 1,
    title: __filename,
    method: 'GET',
    description: 'Return scores from user for a specified type',
    params: [
        {
            type: 'string',
            name: 'user',
            optional: false,
            description: 'id of the user',
        },
        {
            type: 'string',
            name: 'mode',
            optional: true,
            description: '\`\`\`osu\`\`\` or \`\`\`fruits\`\`\` or \`\`\`mania\`\`\` or \`\`\`taiko\`\`\`',
        },
        {
            type: 'string',
            name: 'type',
            optional: true,
            description: '\`\`\`u\`\`\` is a user_id or a username. Use string for usernames or \`\`\`id\`\`\` for user_ids',
        },
        {
            type: 'string',
            name: 'event_days',
            optional: true,
            description: 'Max number of days between now and last event date. Range of 1-31. Optional, default value is 1',
        },
    ],
};
;
const name = (user, type, obj = {}) => __awaiter(void 0, void 0, void 0, function* () {
    const params = {
        u: user,
        m: _mode.indexOf(obj.mode),
        type: obj.type,
        limit: obj.limit,
    };
    if (params.m == -1)
        delete params.m;
    const data = yield (0, request_1.request)(`https://osu.ppy.sh/api/get_user_${type}`, {
        method: 'GET',
        params: params,
    });
    if (data.error)
        return (0, handleErrors_1.handleErrors)(new Error(data.error));
    if (data.length == 0)
        return [];
    if (type == 'best') {
        const format = (0, best_1.default)(data, obj.mode);
        return format;
    }
    ;
    const format = (0, recent_1.default)(data, obj.mode);
    return format;
});
exports.default = name;
