// osu-api-extended@3.1.4/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.v2_matches_detaIls = exports.v2_groups_history = exports.v2_groups_details = exports.v2_forums_topics_details = exports.v2_forums_topics_actions_reply = exports.v2_forums_topics_actions_edit_topic = exports.v2_forums_topics_actions_edit_post = exports.v2_forums_topics_actions_create = exports.v2_comments_list = exports.v2_comments_details = exports.v2_comments_actions_vote = exports.v2_comments_actions_unvote = exports.v2_comments_actions_new = exports.v2_comments_actions_edit = exports.v2_comments_actions_delete = exports.v2_chat_updates = exports.v2_chat_messages = exports.v2_chat_list = exports.v2_chat_details = exports.v2_chat_channels_list = exports.v2_chat_actions_send = exports.v2_chat_actions_new = exports.v2_chat_actions_keepalive = exports.v2_chat_actions_join = exports.v2_changelogs_list_lookup = exports.v2_changelogs_list_all = exports.v2_changelogs_details = exports.v2_beatmaps_packs_list = exports.v2_beatmaps_packs_details = exports.v2_beatmaps_lookup_set = exports.v2_beatmaps_lookup_difficulty = exports.v2_beatmaps_lookup_difficulties = exports.v2_beatmaps_lookup_attributes = exports.v2_beatmaps_events_list = exports.v2_beatmaps_download = exports.v2_beatmaps_discussions_votes = exports.v2_beatmaps_discussions_posts = exports.v2_beatmaps_discussions_list = exports.v2_beatmaps_details_set = exports.v2_beatmaps_details_difficulty = exports.v2_assets_datafiles = exports.v2_assets_backgrounds = exports.download = exports.request = exports.enums = exports.tools = exports.v1 = exports.v2 = exports.callbacks = exports.auth = void 0;
exports.v2_wiki_details = exports.v2_users_lookup = exports.v2_users_list = exports.v2_users_kudosu = exports.v2_users_events = exports.v2_users_details = exports.v2_users_beatmaps = exports.v2_users_activity = exports.v2_spotlights_list = exports.v2_search_all = exports.v2_scores_user_beamap = exports.v2_scores_list_user_recent = exports.v2_scores_list_user_pinned = exports.v2_scores_list_user_firsts = exports.v2_scores_list_user_best = exports.v2_scores_list_leaderboard = exports.v2_scores_list_beatmap_best = exports.v2_scores_list_beatmap_all = exports.v2_scores_download = exports.v2_scores_details = exports.v2_rooms_scores_user_highest = exports.v2_rooms_scores_single = exports.v2_rooms_scores_all = exports.v2_rooms_list = exports.v2_rooms_leaderboard = exports.v2_rooms_details = exports.v2_ranking_list_score = exports.v2_ranking_list_performance = exports.v2_ranking_list_kudosu = exports.v2_ranking_list_country = exports.v2_ranking_list_charts = exports.v2_notifications_list = exports.v2_news_list = exports.v2_news_details = exports.v2_me_friends = exports.v2_me_download_quota = exports.v2_me_details = exports.v2_matches_list = void 0;
exports.auth = __importStar(require("./utility/auth"));
var listeners_1 = require("./utility/listeners");
Object.defineProperty(exports, "callbacks", { enumerable: true, get: function () { return listeners_1.callbacks; } });
exports.v2 = __importStar(require("./routes/v2"));
exports.v1 = __importStar(require("./routes/v1"));
exports.tools = __importStar(require("./tools/index"));
exports.enums = __importStar(require("./types/enums"));
var request_1 = require("./utility/request");
Object.defineProperty(exports, "request", { enumerable: true, get: function () { return request_1.request; } });
Object.defineProperty(exports, "download", { enumerable: true, get: function () { return request_1.download; } });
__exportStar(require("./types/mods"), exports);
__exportStar(require("./types/enums"), exports);
__exportStar(require("./types/tools"), exports);
__exportStar(require("./types/index"), exports);
exports.v2_assets_backgrounds = __importStar(require("./types/v2/assets_backgrounds"));
exports.v2_assets_datafiles = __importStar(require("./types/v2/assets_datafiles"));
exports.v2_beatmaps_details_difficulty = __importStar(require("./types/v2/beatmaps_details_difficulty"));
exports.v2_beatmaps_details_set = __importStar(require("./types/v2/beatmaps_details_set"));
exports.v2_beatmaps_discussions_list = __importStar(require("./types/v2/beatmaps_discussions_list"));
exports.v2_beatmaps_discussions_posts = __importStar(require("./types/v2/beatmaps_discussions_posts"));
exports.v2_beatmaps_discussions_votes = __importStar(require("./types/v2/beatmaps_discussions_votes"));
exports.v2_beatmaps_download = __importStar(require("./types/v2/beatmaps_download"));
exports.v2_beatmaps_events_list = __importStar(require("./types/v2/beatmaps_events_list"));
exports.v2_beatmaps_lookup_attributes = __importStar(require("./types/v2/beatmaps_lookup_attributes"));
exports.v2_beatmaps_lookup_difficulties = __importStar(require("./types/v2/beatmaps_lookup_difficulties"));
exports.v2_beatmaps_lookup_difficulty = __importStar(require("./types/v2/beatmaps_lookup_difficulty"));
exports.v2_beatmaps_lookup_set = __importStar(require("./types/v2/beatmaps_lookup_set"));
exports.v2_beatmaps_packs_details = __importStar(require("./types/v2/beatmaps_packs_details"));
exports.v2_beatmaps_packs_list = __importStar(require("./types/v2/beatmaps_packs_list"));
exports.v2_changelogs_details = __importStar(require("./types/v2/changelogs_details"));
exports.v2_changelogs_list_all = __importStar(require("./types/v2/changelogs_list_all"));
exports.v2_changelogs_list_lookup = __importStar(require("./types/v2/changelogs_list_lookup"));
exports.v2_chat_actions_join = __importStar(require("./types/v2/chat_actions_join"));
exports.v2_chat_actions_keepalive = __importStar(require("./types/v2/chat_actions_keepalive"));
exports.v2_chat_actions_new = __importStar(require("./types/v2/chat_actions_new"));
exports.v2_chat_actions_send = __importStar(require("./types/v2/chat_actions_send"));
exports.v2_chat_channels_list = __importStar(require("./types/v2/chat_channels_list"));
exports.v2_chat_details = __importStar(require("./types/v2/chat_details"));
exports.v2_chat_list = __importStar(require("./types/v2/chat_list"));
exports.v2_chat_messages = __importStar(require("./types/v2/chat_messages"));
exports.v2_chat_updates = __importStar(require("./types/v2/chat_updates"));
exports.v2_comments_actions_delete = __importStar(require("./types/v2/comments_actions_delete"));
exports.v2_comments_actions_edit = __importStar(require("./types/v2/comments_actions_edit"));
exports.v2_comments_actions_new = __importStar(require("./types/v2/comments_actions_new"));
exports.v2_comments_actions_unvote = __importStar(require("./types/v2/comments_actions_unvote"));
exports.v2_comments_actions_vote = __importStar(require("./types/v2/comments_actions_vote"));
exports.v2_comments_details = __importStar(require("./types/v2/comments_details"));
exports.v2_comments_list = __importStar(require("./types/v2/comments_list"));
exports.v2_forums_topics_actions_create = __importStar(require("./types/v2/forums_topics_actions_create"));
exports.v2_forums_topics_actions_edit_post = __importStar(require("./types/v2/forums_topics_actions_edit_post"));
exports.v2_forums_topics_actions_edit_topic = __importStar(require("./types/v2/forums_topics_actions_edit_topic"));
exports.v2_forums_topics_actions_reply = __importStar(require("./types/v2/forums_topics_actions_reply"));
exports.v2_forums_topics_details = __importStar(require("./types/v2/forums_topics_details"));
exports.v2_groups_details = __importStar(require("./types/v2/groups_details"));
exports.v2_groups_history = __importStar(require("./types/v2/groups_history"));
exports.v2_matches_detaIls = __importStar(require("./types/v2/matches_detaIls"));
exports.v2_matches_list = __importStar(require("./types/v2/matches_list"));
exports.v2_me_details = __importStar(require("./types/v2/me_details"));
exports.v2_me_download_quota = __importStar(require("./types/v2/me_download_quota"));
exports.v2_me_friends = __importStar(require("./types/v2/me_friends"));
exports.v2_news_details = __importStar(require("./types/v2/news_details"));
exports.v2_news_list = __importStar(require("./types/v2/news_list"));
exports.v2_notifications_list = __importStar(require("./types/v2/notifications_list"));
exports.v2_ranking_list_charts = __importStar(require("./types/v2/ranking_list_charts"));
exports.v2_ranking_list_country = __importStar(require("./types/v2/ranking_list_country"));
exports.v2_ranking_list_kudosu = __importStar(require("./types/v2/ranking_list_kudosu"));
exports.v2_ranking_list_performance = __importStar(require("./types/v2/ranking_list_performance"));
exports.v2_ranking_list_score = __importStar(require("./types/v2/ranking_list_score"));
exports.v2_rooms_details = __importStar(require("./types/v2/rooms_details"));
exports.v2_rooms_leaderboard = __importStar(require("./types/v2/rooms_leaderboard"));
exports.v2_rooms_list = __importStar(require("./types/v2/rooms_list"));
exports.v2_rooms_scores_all = __importStar(require("./types/v2/rooms_scores_all"));
exports.v2_rooms_scores_single = __importStar(require("./types/v2/rooms_scores_single"));
exports.v2_rooms_scores_user_highest = __importStar(require("./types/v2/rooms_scores_user_highest"));
exports.v2_scores_details = __importStar(require("./types/v2/scores_details"));
exports.v2_scores_download = __importStar(require("./types/v2/scores_download"));
exports.v2_scores_list_beatmap_all = __importStar(require("./types/v2/scores_list_user_beatmap_all"));
exports.v2_scores_list_beatmap_best = __importStar(require("./types/v2/scores_list_user_beatmap_best"));
exports.v2_scores_list_leaderboard = __importStar(require("./types/v2/scores_list_leaderboard"));
exports.v2_scores_list_user_best = __importStar(require("./types/v2/scores_list_user_best"));
exports.v2_scores_list_user_firsts = __importStar(require("./types/v2/scores_list_user_firsts"));
exports.v2_scores_list_user_pinned = __importStar(require("./types/v2/scores_list_user_pinned"));
exports.v2_scores_list_user_recent = __importStar(require("./types/v2/scores_list_user_recent"));
exports.v2_scores_user_beamap = __importStar(require("./types/v2/scores_user_beamap"));
exports.v2_search_all = __importStar(require("./types/v2/search_all"));
exports.v2_spotlights_list = __importStar(require("./types/v2/spotlights_list"));
exports.v2_users_activity = __importStar(require("./types/v2/users_activity"));
exports.v2_users_beatmaps = __importStar(require("./types/v2/users_beatmaps"));
exports.v2_users_details = __importStar(require("./types/v2/users_details"));
exports.v2_users_events = __importStar(require("./types/v2/users_events"));
exports.v2_users_kudosu = __importStar(require("./types/v2/users_kudosu"));
exports.v2_users_list = __importStar(require("./types/v2/users_list"));
exports.v2_users_lookup = __importStar(require("./types/v2/users_lookup"));
exports.v2_wiki_details = __importStar(require("./types/v2/wiki_details"));
