// osu-api-extended@3.1.4/dist/api/v2/comments_list.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.comments_list = void 0;
const handleErrors_1 = require("../../utility/handleErrors");
const request_1 = require("../../utility/request");
const comments_list = (params, addons) => __awaiter(void 0, void 0, void 0, function* () {
    var _a, _b;
    const data = yield (0, request_1.request)(`https://osu.ppy.sh/api/v2/comments`, {
        method: 'GET',
        params: {
            after: params.after_id,
            commentable_type: params.type,
            commentable_id: params.id,
            'cursor[id]': (_a = params.cursor) === null || _a === void 0 ? void 0 : _a.id,
            'cursor[created_at]': (_b = params.cursor) === null || _b === void 0 ? void 0 : _b.created_at,
            parent_id: params.parent_id,
            sort: params.sort,
        },
        addons,
    });
    if (data.error)
        return (0, handleErrors_1.handleErrors)(new Error(data.error));
    return data;
});
exports.comments_list = comments_list;
